### PR TITLE
New MAD-X Parser

### LINKED ISSUE #104: MAD-X input reader: constants, elements, lines, segments
We want to support reading accelerator lattices directly from MAD-X input files.
For this, we will need to support at least the following syntax elements:
- constants
- elements
- lines
- segments
- simple math

from MAD-X definitions.

For that, we could explore if @janfschmidt's `palattice` could be used as a library :)
https://github.com/janfschmidt/palattice

References on MAD-X:
- https://mad.web.cern.ch/mad/webguide/manual.html
- https://github.com/MethodicalAcceleratorDesign/MAD-X/blob/master/doc/latexuguide/format.tex
- MAD websites
  - MAD-X      http://cern.ch/madx   (current version, https://github.com/MethodicalAcceleratorDesign/MAD-X/blob/master/src/mad_parse.c)
  - MAD-9      http://cern.ch/mad9
  - MAD-8      http://cern.ch/mad8
  - MAD-NG   [http://cern.ch/madng/dev](http://cern.ch/mad/releases/madng) (next generation under development)

Example files for MAD-X lattices:
- LBNL/ImpactX/ABLASTR team: https://drive.google.com/drive/u/0/folders/1-BZysj84iEjhDOw-JsEDuu7IytGdl8YN

### LINKED ISSUE #512: MAD-X: `Call, FILENAME` Support
Many MAD project files use nested files, e.g.:
- https://github.com/slaclab/facet2-lattice/blob/master/MAD/FACET2e.mad8

Supporting the MAD-8
```
  CALL, FILENAME="other_file"
```
syntax could be helpful to avoid that we need to linearize such input decks.
http://mad8.web.cern.ch/mad8/user/switch_file.html#call

Seems to be in MAD-9/MAD-X of the syntax
```
  CALL, FILE="other_file"
```
http://mad9.web.cern.ch/mad9/user/node69.html
http://madx.web.cern.ch/madx/webguide/manual.html

### LINKED ISSUE #709: MAD-X Reader: RBEND
Add support for MAD-X [RBEND](https://mad.web.cern.ch/mad/webguide/manual.html#Ch11.S3.I1.ix2) elements in our [src/python/impactx/MADXParser.py](https://github.com/ECP-WarpX/impactx/blob/development/src/python/impactx/MADXParser.py) (this is used by [load_file](https://impactx.readthedocs.io/en/latest/usage/python.html#impactx.elements.KnownElementsList.load_file)).

An `RBEND` will translate to an ImpactX `DipEdge`, `SBend`, `DipEdge`.

### PR DESCRIPTION
**Pull request by ax3l**
## Summary

This PR overhauls two files:
- `MADXParser.py`: a tokenizer and parser that tries to truthfully reimplement the MAD-X C parser, ultimately returning a flattened MAD-X lattice. This file is best written with a powerful LLM (mostly Claude 4.6 was used). It is covered by unit tests for central MAD-X input language features and should not require much updates in the future.
- `madx_to_impactx.py`: the actual translator from MAD-X element to ImpactX element(s). This file we need to be able to maintain "by hand" as we add new elements and handle physics translation corner-cases.

## To Do

- [x] sudo Vibe me a MAD-X parser in Python
  - [x] iterate, test, iterate, test, ...
- [x] Existing MAD-X examples still work.
- [x] Do more complex examples parse now?
  - [x] [simple booster](https://github.com/egstern/impactx/blob/parse_madx_lattice/examples/parse_madx_lattice/sbbooster-cooked.madx)
  - [x] Nov 2024 Booster file `ring_seq_bb_spch_thin (2).madx` (GDrive)
  - [x] Eric's pretty complete `pip-ii-booster-combined.madx` and `sbbooster-cooked.madx` #1271
  - [x] IOTA v8.3 / v84_2 (2019-01-01)
  - [x] LHC: `acc-models-lhc/scenarios/cycle/pp/ramp/1178/model.madx` ([gitlab](https://gitlab.cern.ch/acc-models/acc-models-lhc))
  - [x] FCC: via Arianna
- [ ] How do we want to control element models (linear, paraxial, hamiltonian)?
  - [ ] for MAD-X and PALS as a start, we could have a model option in `from_file(...)`
- [x] Tests: MAD-X input language
- [ ] Tests: commit more MAD-X physics files to our repo?
- [x] Close #104: MAD-X input reader: constants, elements, lines, segments
- [x] Close #512: MAD-X: Call, FILENAME Support
- [x] Close #709: MAD-X Reader: RBEND

### CODE REVIEW

File: src/python/impactx/MADXParser.py [hunk 1]
```diff
@@ -1,644 +1,2287 @@
...
-#   Oct 13th, 2017: original version for pyAcceLEGOrator elements
-#   Aug 10th, 2022: adapted for standalone use
-#
 # -*- coding: utf-8 -*-
+"""
+A MAD-X parser that aims to be compliant with the upstream MAD-X format.
+
+References:
+    - https://mad.web.cern.ch/mad/webguide/manual.html
+    - https://github.com/MethodicalAcceleratorDesign/MAD-X/blob/master/doc/latexuguide/format.tex
```
Comment (ax3l) [inline src/python/impactx/MADXParser.py:13]:
```suggestion
    - https://github.com/MethodicalAcceleratorDesign/MAD-X/blob/master/doc/latexuguide/format.tex
    - https://github.com/MethodicalAcceleratorDesign/MAD-X/blob/master/src/mad_parse.c (and related files)
```

**Review by cemitch99 — COMMENTED** [review]
I think we should add something about MAD-X parsing to the documentation.  However, this should be done with a strong disclaimer.  This feature needs extensive testing before users begin to try it, and blame us if it does not work as expected...

Is there a reason that there are (apparently) unrelated changes to all of the example inputs (e.g., renaming the monitor elements to m1)?

**Comment by ax3l** [comment]
> I think we should add something about MAD-X parsing to the documentation. However, this should be done with a strong disclaimer. This feature needs extensive testing before users begin to try it, and blame us if it does not work as expected...

I will add this and also to PALS loading. Both file loaders are fully in-development.

> Is there a reason that there are (apparently) unrelated changes to all of the example inputs (e.g., renaming the monitor elements to m1)?

Yes, the old MAD-X loader we wrote wrongly ignored the LHS/names of elements. Now we read it properly. That meant I either adjust the analysis scripts of the inputs, because the `monitor.h5` file name is derived on how we call the individual `BeamMonitor(name="...")`, or adjust the inputs.

**Comment by ax3l** [comment]
@cemitch99 I added the warning to `load_file` to make sure people know its status:
https://impactx--1295.org.readthedocs.build/en/1295/usage/python.html#impactx.elements.KnownElementsList.load_file

**Comment by ax3l** [comment]
Discussed today: will also add a runtime warning when `load_file` is used.

**Comment by ax3l** [comment]
@cemitch99 I added runtime warnings when `load_file` is used:
> ...
> examples/chicane/run_chicane_madx.py:28: RuntimeWarning: Our MAD-X parser is under active development and provided as a preview. Please check any loaded MAD-X beams very carefully. Please report your experience and bugs on our issue tracker: https://github.com/BLAST-ImpactX/impactx/issues
  ref = sim.particle_container().ref_particle().load_file("chicane.madx")
> 
> ...
> 
> examples/chicane/run_chicane_madx.py:45: RuntimeWarning: Our MAD-X and PALS parsers are under active development and provided as a preview. Please check any loaded lattice files very carefully. Please report your experience and bugs on our issue tracker: https://github.com/BLAST-ImpactX/impactx/issues
  sim.lattice.load_file("chicane.madx", nslice=25)
> ...

**Comment by ax3l** [comment]
Discussed in last two week's developer meeting: approved & ready for merge :)

Will remove the one little merge conflict in `run_solenoid.py` that came up.

**Comment by ax3l** [comment]
The pyi stub generator shows some bug after this merge, I'll fix that upstream in the next day
